### PR TITLE
Make MultivariateTimeSeries a non-template class

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
1st step to make `GraphTrack` not templated based on the data series `Dimension`.

Bug: #4534
Test: unit tests

### CODE REVIEW

File: src/OrbitGl/GraphTrack.cpp [hunk 1]
```diff
@@ -136,7 +136,7 @@ float GraphTrack<Dimension>::GetLabelYFromValues(
 template <size_t Dimension>
 std::string GraphTrack<Dimension>::GetLabelTextFromValues(
     const std::array<double, Dimension>& values) const {
-  const std::array<std::string, Dimension>& series_names = series_.GetSeriesNames();
+  const std::vector<std::string>& series_names = series_.GetSeriesNames();
   std::optional<uint8_t> value_decimal_digits = series_.GetValueDecimalDigits();
   std::string value_unit = series_.GetValueUnit();
   std::string text;
```
Comment (beckerhe) [inline src/OrbitGl/GraphTrack.cpp:142]: I can add this comment to the proper line, that's why it's here:

Below there are two calls to `series_names[i]`. Can you replace that by `series_names.at(i)` so that it is checked whether the i-th element is still part of the vector.

File: src/OrbitGl/GraphTrack.cpp [hunk 2]
```diff
@@ -175,8 +175,11 @@ void GraphTrack<Dimension>::DrawMouseLabel(PrimitiveAssembler& primitive_assembl
   const Color transparent_white(255, 255, 255, 180);
 
   uint64_t current_mouse_time_ns = draw_context.current_mouse_tick.value();
-  const std::array<double, Dimension>& values =
-      series_.GetPreviousOrFirstEntry(current_mouse_time_ns);
+
+  // TODO(vickyliu): remove this vector to array conversion after changing to not use span.
+  const std::vector<double>& values_tmp = series_.GetPreviousOrFirstEntry(current_mouse_time_ns);
+  std::array<double, Dimension> values;
+  std::copy_n(values_tmp.begin(), Dimension, values.begin());
```
Comment (beckerhe) [inline src/OrbitGl/GraphTrack.cpp:182]: Can you also add a check that ensures that `values_tmp.size() == Dimension`?

File: src/OrbitGl/GraphTrack.cpp [hunk 3]
```diff
@@ -281,7 +284,7 @@ void GraphTrack<Dimension>::DrawSeries(PrimitiveAssembler& primitive_assembler,
   // We skip the last element because we can't calculate time passed between last element
   // and the next one.
   while (current_it != last_it) {
-    std::array<double, Dimension> cumulative_values{current_it->second};
+    std::vector<double> cumulative_values = current_it->second;
```
Comment (beckerhe) [inline src/OrbitGl/GraphTrack.cpp:287]: Same here, can you add a check that verifies the size of the vector is the same as the dimension?

File: src/OrbitGl/MultivariateTimeSeries.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+uint64_t MultivariateTimeSeries::EndTimeInNs() const {
+  absl::MutexLock lock(&mutex_);
+  ORBIT_CHECK(!time_to_series_values_.empty());
+  return time_to_series_values_.rbegin()->first;
+}
+
+const std::vector<double>& MultivariateTimeSeries::GetPreviousOrFirstEntry(uint64_t time) const {
+  absl::MutexLock lock(&mutex_);
+  return GetPreviousOrFirstEntryIterator(time)->second;
```
Comment (beckerhe) [inline src/OrbitGl/MultivariateTimeSeries.cpp:53]: You're returning a reference to some mutex-protected data. This doesn't make a lot of sense, because now the caller has a pointer to the mutex-protected data, but doesn't know that the data should be mutex protected.

Could you instead return a copy? Or is this performance relevant?

File: src/OrbitGl/include/OrbitGl/MultivariateTimeSeries.h [hunk 5]
```diff
@@ -7,140 +7,62 @@
...
-  absl::btree_map<uint64_t, std::array<double, Dimension>> time_to_series_values_
-      ABSL_GUARDED_BY(mutex_);
+  absl::btree_map<uint64_t, std::vector<double>> time_to_series_values_ ABSL_GUARDED_BY(mutex_);
   double min_ ABSL_GUARDED_BY(mutex_) = std::numeric_limits<double>::max();
   double max_ ABSL_GUARDED_BY(mutex_) = std::numeric_limits<double>::lowest();
 
-  const std::array<std::string, Dimension> series_names_;
+  const std::vector<std::string> series_names_;
   const uint8_t value_decimal_digits_;
   const std::string value_unit_;
```
Comment (beckerhe) [inline src/OrbitGl/include/OrbitGl/MultivariateTimeSeries.h:63]: Can you remove the `const`  keywords? Members in classes shouldn't be marked const because it disables move-construction.

File: src/OrbitGl/include/OrbitGl/MultivariateTimeSeries.h [hunk 6]
```diff
@@ -7,140 +7,62 @@
...
-    std::vector<std::pair<uint64_t, std::array<double, Dimension>>> result;
-    result.push_back(*current_it);
-    do {
-      ++current_it;
-      result.push_back(*current_it);
-    } while (current_it != last_iterator);
-
-    return result;
-  }
+      typename absl::btree_map<uint64_t, std::vector<double>>::const_iterator;
```
Comment (beckerhe) [inline src/OrbitGl/include/OrbitGl/MultivariateTimeSeries.h:47]: You won't need the `typename` keyword anymore.

File: src/OrbitGl/MultivariateTimeSeries.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+#include "OrbitGl/MultivariateTimeSeries.h"
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_gl {
+
+MultivariateTimeSeries::MultivariateTimeSeries(absl::Span<const std::string> series_names,
+                                               uint8_t value_decimal_digits, std::string value_unit)
+    : series_names_(series_names.begin(), series_names.end()),
```
Comment (beckerhe) [inline src/OrbitGl/MultivariateTimeSeries.cpp:13]: Instead of using an absl::Span` for the series_names it would be more idiomatic to use a `std::vector<std::string>` and then move it into `series_names_`.

Rule of thumb: Only use `Span<const T>` as a replacement for `const std::vector<T>&` - otherwise getting the series names by value und moving is the better option.

File: src/OrbitGl/LineGraphTrack.cpp [hunk 8]
```diff
@@ -24,7 +24,7 @@ namespace orbit_gl {
 
 template <size_t Dimension>
 [[nodiscard]] static std::array<float, Dimension> GetNormalizedValues(
-    const std::array<double, Dimension>& values, double min, double inverse_value_range) {
+    absl::Span<const double> values, double min, double inverse_value_range) {
```
Comment (beckerhe) [inline src/OrbitGl/LineGraphTrack.cpp:27]: Nit: All the changes here in LineGraphTrack seem to be unrelated for the current PR. Would you mind moving them into the next PR (or whenever you remove Dimension from LineGraphTrack)?

File: src/OrbitGl/BasicPageFaultsTrack.cpp [hunk 9]
```diff
@@ -49,7 +48,7 @@ BasicPageFaultsTrack::BasicPageFaultsTrack(Track* parent,
 void BasicPageFaultsTrack::AddValues(
     uint64_t timestamp_ns, const std::array<double, kBasicPageFaultsTrackDimension>& values) {
   if (previous_time_and_values_.has_value()) {
-    std::array<double, kBasicPageFaultsTrackDimension> differences{};
+    absl::Span<double> differences;
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply src/OrbitGl/BasicPageFaultsTrack.cpp:51]: (Also, this seems to be unrelated to the current change, right?)

[the remark this answers is not in the document]
Comment (vickyliu-go4it) [reply src/OrbitGl/BasicPageFaultsTrack.cpp:51]: Ah, this was a mistake with my previous version that using span to store data. Thanks for pointing this out!

File: src/OrbitGl/GraphTrack.cpp (hunk 3 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/GraphTrack.cpp:287]: I will add the check outside the while loop as all the value vectors in `entries` have the same dimension: this is guaranteed by the check in `series_.AddValue`.

Thanks for the suggestion!

File: src/OrbitGl/LineGraphTrack.cpp (hunk 8 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/LineGraphTrack.cpp:27]: Changed accordingly. Thanks for the suggestion!

This was for the purpose of avoiding the vector to array conversion in `LineGraphTrack`.

**Review by beckerhe — APPROVED** [review]
Thanks Vicky!
